### ISSUE TITLE
Restrict openWindow() to http(s) schemes?

### ISSUE THREAD

**Issue by jakearchibald** [w3c/ServiceWorker#699]
We currently reject on `about:blank` (due to #696), but is there any benefit in allowing schemes other than `http` and `https`?

What would it mean to `openWindow('mailto:hello@example.com')`? etc

I propose we reject on schemes other than http/https. Any objections?

**Comment by annevk** [comment]
What happens if you hit a redirect to a custom scheme? Or execute some JavaScript to navigate to one?

**Comment by mattto** [comment]
Hey, I just noticed the about:blank spec change.

Chrome currently treats all about:* URLs as "about:blank" in openWindow(). Apparently this mimics the behavior of window.open().  cc @mounirlamouri who implemented this in https://codereview.chromium.org/980383004

By the logic of issue #696, it would seem we should instead reject all about:* instead of opening about:blank. Should the spec codify that?

**Comment by jakearchibald** [comment]
Apologies for flip-flopping, but I'm no longer convinced the changes in #696 were a good idea.

Sure, `clients.openWindow('about:blank')` creates a window that cannot be used by the SW, and is arguably useless, but as @annevk points out, a developer would open a page on their origin that navigates to `about:blank` with JS, or a redirect. Or, they could do `clients.openWindow('/whatever').then(c => c.navigate('about:blank'))` once #681 lands.

Making `openWindow` accept a different set of URLs to `window.open` seems needlessly complicated.

**Comment by annevk** [comment]
It seems better if they follow the same code path as much as possible, agreed. This requires some special casing though for `about:blank` as bz pointed out in that bug. You want it to inherit the origin of the service worker (and likely the CSP policy, and maybe some other stuff, we should have something for that).

**Comment by jakearchibald** [comment]
F2F: Allowing `about:blank` means figuring out what origin it is, and if it's service-worker controlled. It's more complicated than it seems, and it's a real edge case. Moving to v2, since we can add it later.

**Comment by monica-ch** [comment]
Bug triage:  Closing the issue as no actual case and real benefit from the restriction. This also break things.

**Comment by annevk** [comment]
OP is asking for additional restrictions. That still seems good to consider as I don't think we have answered the questions in this issue. E.g., what happens for `mailto:` or custom URL schemes today.

**Comment by mkruisselbrink** [comment]
I think the bar for additional restrictions would be significantly higher than loosening restrictions. If there are unspecified/underspecified behaviors around non-http(s) URLs addressing that seems like a good thing (and probably warrants a separate issue). But restricting more than the spec and implementations currently restrict (I didn't verify what implementations do) would imo need some evidence that the benefits outweigh the potential for breakage.

**Comment by annevk** [comment]
Yeah agreed. We should probably test some custom schemes and `javascript` URLs in particular. If there are already tests covering those in WPT I suppose this can be closed.

**Comment by monica-ch** [comment]
Discussed in [TPAC 2025](https://docs.google.com/document/d/1fOjXiGUFf_krkfYaWf5drbqzY7CTxn7EcDYr5ZWjzFg/edit?tab=t.0)

1. Do not expand restrictions beyond existing browser behavior.
2. Keep blocking about:*, because it’s interoperable.
3. Follow window.open() semantics for scheme handling.
4. Collect more metrics before considering additional restrictions.
